### ISSUE TITLE
zfs-generic-iscsi Ubuntu 18.04 targetcli error

### ISSUE THREAD

**Issue by ryanmaffit** [democratic-csi/democratic-csi#127]
# Issue

When using the zfs-generic-iscsi driver with Ubuntu 18.04, the iSCSI target does not get created successfully. The zfs volume gets created and the node attempts to connect to the iSCSI target. 

# Logs

targetcli command

```
{
    "level": "verbose",
    "message": "TargetCLI command: sh -c 'echo \"# create target\ncd /iscsi\ncreate iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11\n\n# setup tpg\ncd /iscsi/iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11/tpg1\n\nset attribute authentication=0\nset attribute cache_dynamic_acls=1\nset attribute demo_mode_write_protect=0\nset attribute generate_node_acls=1\n\n\n# create extent\ncd /backstores/block\ncreate pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11 /dev/zvol/kube/kubernetes/pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11\n\n# add extent to target/tpg\ncd /iscsi/iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11/tpg1/luns\ncreate /backstores/block/pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11\" | targetcli'\n",
    "service": "democratic-csi"
}
```

Executing the command directly on the storage server worked without any errors.

```
sh -c 'echo "# create target
> cd /iscsi
> create iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11
>
> # setup tpg
> cd /iscsi/iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11/tpg1
>
> set attribute authentication=0
> set attribute cache_dynamic_acls=1
> set attribute demo_mode_write_protect=0
> set attribute generate_node_acls=1
>
>
> # create extent
> cd /backstores/block
> create pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11 /dev/zvol/kube/kubernetes/pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11
>
> # add extent to target/tpg
> cd /iscsi/iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11/tpg1/luns
> create /backstores/block/pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11" | targetcli'
targetcli shell version 2.1.fb43
Copyright 2011-2013 by Datera, Inc and others.
For help on commands, type 'help'.

/iscsi> /iscsi> /iscsi> Created target iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11.
Created TPG 1.
Global pref auto_add_default_portal=true
Created default portal listening on all IPs (0.0.0.0), port 3260.
/iscsi> /iscsi> /iscsi> /iscsi/iqn.20...6dcbbf11/tpg1> /iscsi/iqn.20...6dcbbf11/tpg1> Parameter authentication is now '0'.
/iscsi/iqn.20...6dcbbf11/tpg1> Parameter cache_dynamic_acls is now '1'.
/iscsi/iqn.20...6dcbbf11/tpg1> Parameter demo_mode_write_protect is now '0'.
/iscsi/iqn.20...6dcbbf11/tpg1> Parameter generate_node_acls is now '1'.
/iscsi/iqn.20...6dcbbf11/tpg1> /iscsi/iqn.20...6dcbbf11/tpg1> /iscsi/iqn.20...6dcbbf11/tpg1> /iscsi/iqn.20...6dcbbf11/tpg1> /backstores/block> Created block storage object pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11 using /dev/zvol/kube/kubernetes/pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11.
/backstores/block> /backstores/block> /backstores/block> /iscsi/iqn.20...f11/tpg1/luns> Created LUN 0.
/iscsi/iqn.20...f11/tpg1/luns> exit
Global pref auto_save_on_exit=true
Last 10 configs saved in /etc/rtslib-fb-target/backup.
Configuration saved to /etc/rtslib-fb-target/saveconfig.json
```

```
# targetcli ls
o- / ......................................................................................................................... [...]
  o- backstores .............................................................................................................. [...]
  | o- block .................................................................................................. [Storage Objects: 1]
  | | o- pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11  [/dev/zvol/kube/kubernetes/pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11 (20.0GiB) write-thru activated]
  | o- fileio ................................................................................................. [Storage Objects: 0]
  | o- pscsi .................................................................................................. [Storage Objects: 0]
  | o- ramdisk ................................................................................................ [Storage Objects: 0]
  o- iscsi ............................................................................................................ [Targets: 1]
  | o- iqn.2003-01.org.linux-iscsi.ubuntu-19.x8664:pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11 ......................................... [TPGs: 1]
  |   o- tpg1 .................................................................................................. [gen-acls, no-auth]
  |     o- acls .......................................................................................................... [ACLs: 0]
  |     o- luns .......................................................................................................... [LUNs: 1]
  |     | o- lun0  [block/pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11 (/dev/zvol/kube/kubernetes/pvc-ded45858-a6b6-49be-8592-e6ce6dcbbf11)]
  |     o- portals .................................................................................................... [Portals: 1]
  |       o- 0.0.0.0:3260 ..................................................................................................... [OK]
  o- loopback ......................................................................................................... [Targets: 0]
  o- vhost ............................................................................................................ [Targets: 0]
```

targetcli  response

```
{
    "level": "verbose",
    "message": "TargetCLI response: {\"stderr\":\"Traceback (most recent call last):\\n  File \\\"/usr/lib/python3/dist-packages/configshell_fb/shell.py\\\", line 893, in run_interactive\\n    old_completer = readline.get_completer()\\nNameError: name 'readline' is not defined\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \\\"/usr/bin/targetcli\\\", line 121, in <module>\\n    main()\\n  File \\\"/usr/bin/targetcli\\\", line 111, in main\\n    shell.run_interactive()\\n  File \\\"/usr/lib/python3/dist-packages/configshell_fb/shell.py\\\", line 899, in run_interactive\\n    readline.set_completer(old_completer)\\nNameError: name 'readline' is not defined\\n\",\"stdout\":\"targetcli shell version 2.1.fb43\\nCopyright 2011-2013 by Datera, Inc and others.\\nFor help on commands, type 'help'.\\n\\n\",\"code\":1}",
    "service": "democratic-csi"
}
```

```
{
    "stderr": "Traceback (most recent call last):\\n  File \\"/usr/lib/python3/dist-packages/configshell_fb/shell.py\\", line 893, in run_interactive\\n    old_completer = readline.get_completer()\\nNameError: name 'readline' is not defined\\n\\nDuring handling of the above exception, another exception occurred:\\n\\nTraceback (most recent call last):\\n  File \\"/usr/bin/targetcli\\", line 121, in <module>\\n    main()\\n  File \\"/usr/bin/targetcli\\", line 111, in main\\n    shell.run_interactive()\\n  File \\"/usr/lib/python3/dist-packages/configshell_fb/shell.py\\", line 899, in run_interactive\\n    readline.set_completer(old_completer)\\nNameError: name 'readline' is not defined\\n",
    "stdout": "targetcli shell version 2.1.fb43\\nCopyright 2011-2013 by Datera, Inc and others.\\nFor help on commands, type 'help'.\\n\\n",
    "code": 1
}
```

```
Traceback (most recent call last):
  File \\"/usr/lib/python3/dist-packages/configshell_fb/shell.py\\", line 893, in run_interactive
    old_completer = readline.get_completer()
NameError: name 'readline' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File \\"/usr/bin/targetcli\\", line 121, in <module>
    main()
  File \\"/usr/bin/targetcli\\", line 111, in main
    shell.run_interactive()
  File \\"/usr/lib/python3/dist-packages/configshell_fb/shell.py\\", line 899, in run_interactive
    readline.set_completer(old_completer)
NameError: name 'readline' is not defined
```

# Environment

```
# targetcli --version
/usr/bin/targetcli version 2.1.fb43

# dpkg -l | grep target
ii  python-rtslib-fb                 2.1.57+debian-4                           all          object API for managing the Linux LIO kernel target - Python 2.x
ii  python3-rtslib-fb                2.1.57+debian-4                           all          object API for managing the Linux LIO kernel target - Python 3.x
ii  targetcli-fb                     2.1.43-1                                  all          Command shell for managing the Linux LIO kernel target
```

# Workaround

I found a [lanchpad bug report](https://bugs.launchpad.net/ubuntu/+source/python-configshell-fb/+bug/1776761) with the same error indicating it is a configshell-fb bug that may be the root cause of the problem. 

After upgrading the storage server to Ubuntu 20.04 using `do-release-upgrade` the targetcli commands succeeded in setting up the iSCSI target and everything works as expected.

**Comment by travisghansen** [comment]
OK, I've got a 'fix' in for this: https://github.com/democratic-csi/democratic-csi/commit/3ab402367804ac05480bae2d5ee7f74aa2a041b5

I'll make s new snapshot in the next day or so that will include the fix so it will be 'official' and all. If you wish to test beforehand you can try the images with the `next` tag.

**Comment by ryanmaffit** [comment]
@travisghansen Thanks, I'll give it a shot this week and report back.

**Comment by travisghansen** [comment]
Released in `v1.4.3`.
